autogo/def.bzl

```diff
@@ -0,0 +1,99 @@
+    return linux
+  fail("autogo only supports linux and mac/darwin, not %s" % repo_ctx.os.name)
+
+
+def _impl(repo_ctx):
+  print("Generating rules for %s..." % repo_ctx.attr.prefix)
+  # tell gazelle which package it is generating
+  prefix = repo_ctx.attr.prefix
+  # find the autogo binary for the host env
+  autogo = repo_ctx.path(_select(repo_ctx, repo_ctx.attr._autogo_linux, repo_ctx.attr._autogo_darwin))
+  # find the gazelle binary for the host env
+  gazelle = repo_ctx.path(_select(repo_ctx, repo_ctx.attr._gazelle_linux, repo_ctx.attr._gazelle_darwin))
+  # get the path to the root workspace
+  root_workspace = repo_ctx.path(repo_ctx.attr._workspace).dirname
+
+  # Create a shadow copy of root_workspace in the repo workspace
+
+  # First, delete anything we created last time
+  # TODO(fejta): massively unoptimized, but expedient way to ensure correctness
+  _exec(repo_ctx, ["rm", "-rf", repo_ctx.path(".")])
+  _exec(repo_ctx, ["touch", repo_ctx.path("BUILD.bazel")]) # required by bazel
+  # Update the timestamp of the autogo binary, invalidating the autogo result cache.
+  # Critically, this ensures that we reevaluate the root workspace for any changes.
+  # Otherwise, if a file/package is added/deleted/moved we will not notice and not rerun gazelle.
+  _exec(repo_ctx, ["touch", autogo])
+  # Run the autogo binary (see shadow.go), which clones the dir tree and symlinks in relevant files
+  _exec(repo_ctx, [autogo, root_workspace, repo_ctx.path(".")])
+  # Run gazelle to auto-create rules
+  _exec(repo_ctx,
+    [gazelle, "--repo_root", repo_ctx.path(""), "--go_prefix", repo_ctx.attr.prefix])
+
+_autogo_generate = repository_rule(
+    implementation=_impl,
+    local=True,
+    attrs={
+        "prefix": attr.string(mandatory=True),
+        "_workspace": attr.label(
+            allow_single_file=True,
+            default="@//:WORKSPACE"),
+        "_autogo_linux": attr.label(
+            allow_single_file=True,
+            default="//autogo:bin/linux_amd64/autogo"),
+        "_autogo_darwin": attr.label(
+            allow_single_file=True,
+            default="//autogo:bin/darwin_amd64/autogo"),
+        "_gazelle_linux": attr.label(
+            allow_single_file=True,
+            default="//autogo:bin/linux_amd64/gazelle"),
+        "_gazelle_darwin": attr.label(
+            allow_single_file=True,
+            default="//autogo:bin/darwin_amd64/gazelle"),
+        },
+)
+
+def autogo_generate(name="autogo", **kw):
+  """Generates automatic rules for go packages at @name (usually @autogo).
+
+  Usage:
+    autogo_generate(name="autogo", prefix="github.com/my/go/get/path")
+  """
+  go_rules_dependencies()
+  go_register_toolchains()
+  _autogo_generate(name=name, **kw)
+
+def _exec(repo_ctx, cmd, quiet=True, **kw):
+  """Run a command and fail if it returns non-zero."""
+  # set quiet=False to help debugging
+  ret = repo_ctx.execute(cmd, quiet=quiet, **kw)
+  if ret.return_code:
+    print(cmd, "returned", ret.return_code)
```

Comment:
should these only be printed if `quiet=False`?

Reply:
Done